File: src/massmatrix.cpp
#include "massmatrix.hpp"

Eigen::SparseMatrix<double>
trimesh_massmatrix(
    Eigen::Ref<const compas::RowMatrixXd> V,
    Eigen::Ref<const compas::RowMatrixXi> F,
    const std::string& type
) {
    Eigen::SparseMatrix<double> M;
    igl::massmatrix(V, F, igl::MASSMATRIX_TYPE_BARYCENTRIC, M);
    return M;
}

NB_MODULE(_massmatrix, m) {

    m.def(
        "trimesh_massmatrix",
        &trimesh_massmatrix,
        "Compute the mass matrix for a triangle mesh.",
        "V"_a, "F"_a, "type"_a="barycentric"
    );
}

File: src/compas_libigl/massmatrix.py
import numpy as np
from compas.plugins import plugin

from compas_libigl import _massmatrix


@plugin(category="trimesh")
def trimesh_massmatrix(M):
    """Compute the mass matrix of a triangle mesh.

    Parameters
    ----------
    M : tuple[list[list[float]], list[list[int]]]
        A mesh represented by a tuple of (vertices, faces)
        where vertices are 3D points and faces are triangles

    Returns
    -------
    scipy.sparse.csc_matrix
        The mass matrix in sparse format.
    """
    V, F = M
    V = np.asarray(V, dtype=np.float64)
    F = np.asarray(F, dtype=np.int32)
    return _massmatrix.trimesh_massmatrix(V, F)
